### ISSUE TITLE
dataplane compatibility with haproxy 2.7.x

### ISSUE THREAD

**Issue by josephkeiser** [haproxytech/dataplaneapi#297]
In my testing of DataPlane API v2.7.2+ with HAProxy v2.7.x I've been unable to successfully commit any changes using the API. GET requests seem to work fine, however, any command that attempts to make changes fails. I've played around with various version combinations, but it seems that DataPlane API is incompatible with HAProxy v2.7.x. Is this known/expected at this time?

**Comment by mjuraga** [comment]
This is not known or expected. In fact we have CI jobs running with HAProxy 2.7 and everything works. What errors are you getting running those requests?

**Comment by josephkeiser** [comment]
This incompatibility was reported in a different issue, and I couldn't find a new one, so I opened this. The previous mentions start here :https://github.com/haproxytech/dataplaneapi/issues/119#issuecomment-1435065284. I am seeing the same behavior reported.

I am running haproxy in docker v2.7.4 (alpine 3.17) on an ubuntu v18 box, along with dataplane api v2.7.2.

```
program api
  command /usr/bin/dataplaneapi --scheme http --host localhost --port 4444 --haproxy-bin /usr/local/sbin/haproxy --config-file /usr/local/etc/haproxy/haproxy.cfg --reload-cmd "kill -SIGUSR2 1" --reload-delay 5 --restart-cmd "kill -SIGUSR2 1" --userlist dataplane_users --write-timeout=120s
  no option start-on-reload
```

A POST to replace the configuration with the same running configuration (`v2/services/haproxy/configuration/raw?force_reload=true&skip_version=true`) causes this response. The transactionId is different for every request.
```
{
    "code": 400,
    "message": "14: err transactionId=d40ce47a-6edd-4a66-9e86-b2e65340fcd1 "
}
```
 The same response is received for attempts to create a backend, specifying just the backends name. In essence, anything that creates a transaction (directly, or indirectly) does not succeed.

**Comment by josephkeiser** [comment]
haproxy:2.7.4-alpine3.17 (https://hub.docker.com/layers/library/haproxy/2.7.4-alpine3.17/images/sha256-232cf245a026a583df64c10a8d6eeb6575162aeaa36d633929c094e00ec90692?context=explore)

**Comment by josephkeiser** [comment]
I didn't know those other images existed. However, I will have an issue if I need to switch to them, given that you do not publish an s390x architecture version.

Also, the version of dataplaneapi that is packaged with 2.7.5 of haproxy-alpine does not start, so I've had to build a new image with an older version (2.7.2) of dataplaneapi to do my testing.

Regardless, I'm still getting the same error. If this is assumed to be a file permissions error, what files/directories is dataplaneapi attempting to manage? Perhaps that's where I should continue debugging.

**Comment by dkorunic** [comment]
We would offer `linux/s390x`, but sadly Qemu s390x emulation doesn't work reliably and tls/x509 code in Golang runtime has numerous issues when trying to compile Dataplaneapi inside the emulation :-( Can you make sure that that dataplaneapi has permissions to write to required folders? If unsure, take a look at our [Dockerfile](https://github.com/haproxytech/haproxy-docker-alpine-quic/blob/main/2.6/Dockerfile#L69), more specific into lines where are `chown` and `chmod` commands?

**Comment by josephkeiser** [comment]
I've looked at the Dockerfile and cannot find anything that jumps out at me as being wrong. Therefore, I did more testing, and simply swapping out the haproxy version from 2.7.5 to 2.6.10, with no other changes, works just fine. Here are the files I'm using to build my image, run haproxy, and run the image. The dataplaneapi file that is being copied in with the Dockerfile is [here (using Linux_x86_64)](https://github.com/haproxytech/dataplaneapi/releases/tag/v2.7.2).

Dockerfile
```
# FROM haproxytech/haproxy-alpine:2.6.10
FROM haproxytech/haproxy-alpine:2.7.5

LABEL Description="Alpine:3.17, HAProxy:2.6.10, Dataplane:2.7.2"

COPY ./dataplaneapi /usr/bin/
RUN chmod +x /usr/bin/dataplaneapi

COPY haproxy.cfg /usr/local/etc/haproxy/
```

HAProxy config
```
global
  user haproxy
  group haproxy
  master-worker
  maxconn 10000
  pidfile /var/run/haproxy.pid
  stats socket /var/run/haproxy.sock level admin mode 660
  log fd@2 local2

defaults
  mode tcp
  timeout connect 30s
  timeout client 180s
  timeout server 180s
  default-server init-addr last,libc,none resolvers dockerdns resolve-prefer ipv4

userlist dataplane_users
  user admin insecure-password password

resolvers dockerdns
  nameserver dns 127.0.0.11:53
  hold nx 5s
  hold obsolete 5s
  hold other 5s
  hold refused 5s
  hold timeout 5s
  hold valid 5s
  timeout resolve 2s
  timeout retry 2s
  resolve_retries 100

program api
  command /usr/bin/dataplaneapi --scheme http --host localhost --port 4444 --haproxy-bin /usr/local/sbin/haproxy --config-file /usr/local/etc/haproxy/haproxy.cfg --reload-cmd "kill -SIGUSR2 1" --reload-delay 5 --restart-cmd "kill -SIGUSR2 1" --userlist dataplane_users --write-timeout=120s --log-to=stdout --log-level=trace
  no option start-on-reload

frontend dataplane
  mode http
  bind :5555
  use_backend dataplane

backend dataplane
  mode http
  http-request set-header Authorization "Basic YWRtaW46cGFzc3dvcmQ="
  server dataplane localhost:4444
```

docker-compose.yml
```
version: '3.7'
services:
  haproxy:
    image: haproxy:test
    volumes:
      - config:/usr/local/etc/haproxy:rw
    ports:
      - target: 5555
        published: 5555
        protocol: tcp
        mode: host
networks:
  default:
    external: true
    name: profiler_network
volumes:
  config:
```

**Comment by mjuraga** [comment]
@bernielomax we are still investigating this, and hopefully we'll have a fix soon, as soon as we fix this we'll be releaseing a new patch version with all the relevant fixes.

**Comment by oliwer** [comment]
I managed to debug this a bit further. The error you see happens when dataplaneapi calls HAProxy to validate the new haproxy.cfg. This particular setup causes a segfault during HAProxy's early initialization:

```
Reading symbols from /usr/local/sbin/haproxy...

[New LWP 84]

warning: Section `.reg-xstate/84' in core file too small.
Core was generated by `/usr/local/sbin/haproxy -f /tmp/haproxy/haproxy.cfg.2a93b74a-39af-4fdd-a695-b75'.
Program terminated with signal SIGSEGV, Segmentation fault.

warning: Section `.reg-xstate/84' in core file too small.
#0  a_crash () at ./arch/x86_64/atomic_arch.h:108
108	./arch/x86_64/atomic_arch.h: No such file or directory.
(gdb) bt
#0  a_crash () at ./arch/x86_64/atomic_arch.h:108
#1  get_meta (p=p@entry=0x7ffe7d7c0f65 "3") at src/malloc/mallocng/meta.h:131
#2  0x00007fe07c5a7931 in __libc_free (p=0x7ffe7d7c0f65) at src/malloc/mallocng/free.c:105
#3  0x00007fe07c5a70f5 in free (p=<optimized out>) at src/malloc/free.c:5
#4  0x000055ba72f94696 in memvprintf (out=0x7ffe7d7bf0b0, format=0x55ba730abc1e "%d", orig_args=orig_args@entry=0x7ffe7d7befd0)
    at src/tools.c:4215
#5  0x000055ba72f947c4 in memprintf (out=out@entry=0x7ffe7d7bf0b0, format=format@entry=0x55ba730abc1e "%d") at src/tools.c:4228
#6  0x000055ba7306c34a in startup_logs_init () at src/errors.c:148
#7  0x000055ba72ec6ea0 in init (argc=4, argv=0x7ffe7d7bf518) at src/haproxy.c:1970
#8  main (argc=4, argv=0x7ffe7d7bf518) at src/haproxy.c:3275
```

The error seems to be related to be related to the master-worker mode. I have not been able to reproduce the crash when running HAProxy manually.

**Comment by wlallemand** [comment]
This was fixed in haproxy upstream with this patch https://github.com/haproxy/haproxy/commit/b4e651f12f90021c7f92f7b31999315206121eb3. But as a fix in the dataplane API, you could cleanup the complete env before starting haproxy -c. (like with `env -i`)

**Comment by oliwer** [comment]
For the record, we cannot simply clean the whole environment because HAProxy needs to expand variables for the validation to pass. So in dataplaneapi, I think we'll just remove the environment variables mentioned in @wlallemand's patch.

**Comment by mjuraga** [comment]
We have a fix for this already, a new release is coming soon.

**Comment by mjuraga** [comment]
Hi, this has been released in https://github.com/haproxytech/dataplaneapi/releases/tag/v2.7.4 can you please check if this works for you?

**Comment by josephkeiser** [comment]
Here's what I've found in my testing. I'm not sure if anyone else has had an opportunity to try it out?

- dataplane API **v2.7.2 works** with HAProxy v2.7.8
- dataplane API **v2.7.3 seg faults on startup**
- dataplane API **v2.7.4 seg faults on startup**
```
[NOTICE]   (1) : New program 'api' (8) forked
[NOTICE]   (1) : New worker (9) forked
[NOTICE]   (1) : Loading success.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0x11038db]

goroutine 1 [running]:
github.com/haproxytech/dataplaneapi.SyncWithFileSettings(0xc000051340, 0xc000050170?)
	github.com/haproxytech/dataplaneapi/dataplaneapi_configuration.go:38 +0xfb
main.startServer(0xc000373900)
	github.com/haproxytech/dataplaneapi/cmd/dataplaneapi/main.go:137 +0x7f2
main.main()
	github.com/haproxytech/dataplaneapi/cmd/dataplaneapi/main.go:59 +0x3d
[NOTICE]   (1) : haproxy version is 2.7.8-58c657f
[ALERT]    (1) : Current program 'api' (8) exited with code 2 (Exit)
[ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
[ALERT]    (1) : Current worker (9) exited with code 143 (Terminated)
[WARNING]  (1) : All workers exited. Exiting... (2)
```

haproxy.cfg:
```
global
  master-worker
  maxconn 10000
  pidfile /var/run/haproxy.pid
  log fd@2 local2

defaults
  mode tcp
  timeout connect 30s
  timeout client 180s
  timeout server 180s
  default-server init-addr last,libc,none resolvers dockerdns resolve-prefer ipv4

userlist dataplane_users
  user admin insecure-password password

resolvers dockerdns
  nameserver dns 127.0.0.11:53
  hold nx 5s
  hold obsolete 5s
  hold other 5s
  hold refused 5s
  hold timeout 5s
  hold valid 5s
  timeout resolve 2s
  timeout retry 2s
  resolve_retries 100

program api
  command /usr/bin/dataplaneapi --scheme http --host localhost --port 4444 --haproxy-bin /usr/local/sbin/haproxy --config-file /usr/local/etc/haproxy/haproxy.cfg --reload-cmd "kill -SIGUSR2 1" --reload-delay 5 --restart-cmd "kill -SIGUSR2 1" --userlist dataplane_users --write-timeout=120s --log-to=stdout --log-level=trace
  no option start-on-reload

frontend dataplane
  mode http
  bind :5555
  use_backend dataplane

backend dataplane
  mode http
  http-request set-header Authorization "Basic YWRtaW46cGFzc3dvcmQ="
  server dataplane localhost:4444
```

run command:
```shell
docker run --rm -v ./haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg haproxytech/haproxy-alpine:2.7.8
```

**Comment by mjuraga** [comment]
Hi @josephkeiser this is another issue, addressed here: https://github.com/haproxytech/dataplaneapi/issues/293

It will be in the coming release `v2.7.5` I missed the backport.

**Comment by mjuraga** [comment]
`v2.7.5` is now released: https://github.com/haproxytech/dataplaneapi/releases/tag/v2.7.5

**Comment by josephkeiser** [comment]
Thanks @mjuraga, everything does appear to be working well again.

Thank you for your work on this.
